subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/configurations/model/ConfigurationReportModelFactory.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.diagnostics.internal.configurations.model;
+
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.ConfigurationVariant;
+import org.gradle.api.artifacts.PublishArtifact;
+import org.gradle.api.attributes.Attribute;
+import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.attributes.AttributeMatchingStrategy;
+import org.gradle.api.attributes.AttributesSchema;
+import org.gradle.api.capabilities.Capability;
+import org.gradle.api.internal.artifacts.configurations.ConfigurationInternal;
+import org.gradle.api.internal.attributes.DefaultCompatibilityRuleChain;
+import org.gradle.api.internal.attributes.DefaultDisambiguationRuleChain;
+import org.gradle.api.internal.file.FileResolver;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * Factory for creating {@link ConfigurationReportModel} instances which represent the configurations present in a project.
+ */
+public final class ConfigurationReportModelFactory {
+    private final FileResolver fileResolver;
+
+    public ConfigurationReportModelFactory(FileResolver fileResolver) {
+        this.fileResolver = fileResolver;
+    }
+
+    public ConfigurationReportModel buildForProject(Project project) {
+        final ConfigurationRuleScanner scanner = new ConfigurationRuleScanner(project);
+        final List<ReportAttribute> attributesWithCompatibilityRules = scanner.getAttributesWithCompatibilityRules();
+        final List<ReportAttribute> attributesWithDisambiguationRules = scanner.getAttributesWithDisambiguationRules();
+
+        final Map<String, ReportConfiguration> convertedConfigurations = new ConcurrentHashMap<>(project.getConfigurations().size());
```

Comment:
I'm not sure why this is a `Concurrent` hash map? It doesn't seem to be written to concurrently.

Reply:
I see it was intended for `computeIfAbsent` below, but that is still incorrect code. This should not need to be a concurrent hash map.

Reply:
It is necessary, otherwise (at least) the tests in `ResolvableConfigurationsReportTaskIntegrationTest` throw `ConcurrentModificationException`s.